player/lua/select.lua

```diff
@@ -353,6 +353,113 @@ mp.add_key_binding(nil, "select-audio-device", function ()
     })
 end)
 
+local history_file_path =
+    mp.command_native({"expand-path", mp.get_property("watch-history-path")})
+
+mp.register_event("file-loaded", function ()
+    if not mp.get_property_native("save-watch-history") then
+        return
+    end
+
+    local history_file, error_message = io.open(history_file_path, "a")
+    if not history_file then
+        show_error("Failed to write the watch history: " .. error_message)
+        return
+    end
+
+    local path = mp.command_native({"normalize-path", mp.get_property("path")})
+    local title = mp.get_property("playlist/" .. mp.get_property("playlist-pos") .. "/title")
+
+    history_file:write(utils.format_json({os.time(), path, title}) .. "\n")
+    history_file:close()
+end)
+
+local function add_history_entry(line, items, paths, seen, osd_playlist_entry)
+    local entry = utils.parse_json(line)
+
+    if not entry then
+        mp.msg.warn(line .. " in " .. history_file_path .. " is not valid JSON.")
+        return
+    end
+
+    local time, path, title = unpack(entry)
+
+    if seen[path] then
+        return
+    end
+    seen[path] = true
+
+    local status, date = pcall(os.date, "(%Y-%m-%d %H:%M) ", time)
+
+    if not status or not path then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
+        return
+    end
+
+    for i, seen_path in ipairs(paths) do
+        if seen_path == path then
+            table.remove(items, i)
+            table.remove(paths, i)
+            break
+        end
+    end
+
+    local item = path
+    if title and osd_playlist_entry == "title" then
+        item = title
+    elseif title and osd_playlist_entry == "both" then
+        item = title .. " (" .. path .. ")"
+    end
+
+    items[#items + 1] = date .. item
+    paths[#paths + 1] = path
+end
+
+mp.add_key_binding(nil, "select-watch-history", function ()
+    local history_file, error_message = io.open(history_file_path)
+    if not history_file then
+        show_warning(mp.get_property_native("save-watch-history")
+                     and error_message
+                     or "Enable --save-watch-history")
+        return
+    end
+
+    local lines = {}
+    local items = {}
+    local paths = {}
+    local seen = {}
+    local osd_playlist_entry = mp.get_property("osd-playlist-entry")
+
+    for line in history_file:lines() do
+        table.insert(lines, line)
+    end
+    history_file:close()
+
+    for i = #lines, 1, -1 do
+        add_history_entry(lines[i], items, paths, seen, osd_playlist_entry)
+    end
+
+    items[#items+1] = "Clear history"
+
+    input.select({
+        prompt = "Select a file:",
+        items = items,
```

Comment:
Food for though. Maybe `input.select` could ingest dicts instead of strings. And have formatter function to convert to string. Would simplify this `items` and  `paths` lists, that are basically duplicated, only because we join string.

Reply:
I think we can replace items with custom object with to string custom conversion. This works well https://0x0.st/8iJx.lua So that we keep only single list of `Items`.